server/project_member.go

```diff
@@ -2,18 +2,177 @@ package server
+				Refresher:    s.refreshToken(ctx, repo.ID),
+			},
+			vcs.InstanceURL,
+			repo.ExternalID,
+		)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch repository member from VCS, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+		}
+
+		// The following block will check whether the relevant principal exists in our system.
+		// If the principal does not exist, we will try to create one out of the vcs member info.
+		createList := make([]*api.ProjectMemberCreate, 0)
+		for _, projectMember := range vcsProjectMemberList {
+			if vcs.Type != projectMember.RoleProvider {
+				return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Invalid role provider, expected: %v, got: %v", vcs.Type, projectMember.RoleProvider)).SetInternal(err)
+			}
+
+			findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+			principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+			}
+			// TODO(zilong): the creation of principal is not within the transaction. Need to move this block to the same transaction of creating & deleting members
+			if principal == nil { // try to create principal
+				signupInfo := &api.Signup{
+					Name:  projectMember.Name,
+					Email: projectMember.Email,
+					// Principal created via this method would have no chance to set their password.
+					// To prevent potential security issues, we use random string to set up her password.
+					// This is another safety measure since we already disallow user login via password
+					// if the principal uses external auth provider
+					Password: common.RandomString(20),
+				}
+				createdPrincipal, httpErr := trySignup(ctx, s, signupInfo, c.Get(getPrincipalIDContextKey()).(int))
+				if httpErr != nil {
+					return httpErr
+				}
+				principal = createdPrincipal
+			}
+
+			providerPayload := &api.ProjectRoleProviderPayload{
+				VCSRole:    projectMember.VCSRole,
+				LastSyncTs: time.Now().UTC().Unix(),
+			}
+			providerPayloadBytes, _ := json.Marshal(providerPayload)
+			createProjectMember := &api.ProjectMemberCreate{
+				ProjectID:    projectID,
+				CreatorID:    c.Get(getPrincipalIDContextKey()).(int),
+				PrincipalID:  principal.ID,
+				Role:         projectMember.Role,
+				RoleProvider: api.ProjectRoleProvider(projectMember.RoleProvider),
+				Payload:      string(providerPayloadBytes),
+			}
+			createList = append(createList, createProjectMember)
+		}
+
+		oldProjectMemberList, err := s.ProjectMemberService.GetSetProjectMember(ctx, projectID, c.Get(getPrincipalIDContextKey()).(int), createList)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to sync project member from VCS").SetInternal(err)
+		}
+
+		// create activity for member deletion
+		for _, projectMember := range oldProjectMemberList {
```

Comment:
What if oldProjectMemberList and createList have overlapping?

We might consider s.ProjectMemberService.GetSetProjectMember returning the diff (newly added members, removed members). Also it might be fine to call it SetProjectMember instead of GetSetProjectMember.